libraries/AP_Beacon/AP_Beacon_Marvelmind.cpp

```diff
@@ -29,81 +44,28 @@ AP_Beacon_Marvelmind::AP_Beacon_Marvelmind(AP_Beacon &frontend, AP_SerialManager
     }
 }
 
-bool AP_Beacon_Marvelmind::get_or_alloc_beacon(struct StationaryBeaconPosition &b, uint8_t address)
+//////////////////////////////////////////////////////////////////////////////
+// Calculate CRC (Modbus) for array of bytes
+// buf: input buffer
+// len: size of buffer
+// returncode: CRC value
+//////////////////////////////////////////////////////////////////////////////
+uint16_t AP_Beacon_Marvelmind::calc_crc_modbus(uint8_t *buf, uint16_t len)
 {
-    const uint8_t n_used = hedge->positions_beacons.num_beacons;
-    if (n_used != 0) {
-        for (uint8_t i = 0; i < n_used; i++) {
-            if (hedge->positions_beacons.beacons[i].address == address) {
-                b = hedge->positions_beacons.beacons[i];
-                return true;
+    uint16_t crc = 0xFFFF;
+    for (uint16_t pos = 0; pos < len; pos++) {
+        crc ^= (uint16_t) buf[pos]; // XOR byte into least sig. byte of crc
+        for (uint8_t i = 8; i != 0; i--) { // Loop over each bit
+            if ((crc & 0x0001) != 0) { // If the LSB is set
+                crc >>= 1; // Shift right and XOR 0xA001
+                crc ^= 0xA001;
```

Comment:
ok, now I feel stupid. It says all this in the function header and name. <facepalm>